From 1be84b20aed68ff76d467324027dfb6d42cb79ce Mon Sep 17 00:00:00 2001
From: Ionut Nechita <ionut_n2001@yahoo.com>
Date: Sun, 3 Apr 2022 18:02:04 +0300
Subject: [PATCH 27/58] SUNLIGHT: mt76: mt7921: start reworking tx rate
 reporting #1

Description:
 - Similar to mt7915, introduce mt7921_txwi_free to report tx rate to
   mac80211. This is a preliminary patch to add proper tx status report.

Change-Id: I373a60c587cfc256f27678455d3fc4f833bc48d7
Signed-off-by: Lorenzo Bianconi <lorenzo@kernel.org>
Signed-off-by: Ionut Nechita <ionut_n2001@yahoo.com>
---
 .../net/wireless/mediatek/mt76/mt7921/mac.c   | 76 ++++++++++---------
 .../wireless/mediatek/mt76/mt7921/mt7921.h    |  2 -
 2 files changed, 42 insertions(+), 36 deletions(-)

diff --git a/drivers/net/wireless/mediatek/mt76/mt7921/mac.c b/drivers/net/wireless/mediatek/mt76/mt7921/mac.c
index c093920a597d..661c51f55464 100644
--- a/drivers/net/wireless/mediatek/mt76/mt7921/mac.c
+++ b/drivers/net/wireless/mediatek/mt76/mt7921/mac.c
@@ -985,7 +985,7 @@ mt7921_tx_check_aggr(struct ieee80211_sta *sta, __le32 *txwi)
 
 static void
 mt7921_tx_complete_status(struct mt76_dev *mdev, struct sk_buff *skb,
-			  struct ieee80211_sta *sta, u8 stat,
+			  struct ieee80211_sta *sta, u8 clean_status,
 			  struct list_head *free_list)
 {
 	struct ieee80211_tx_info *info = IEEE80211_SKB_CB(skb);
@@ -1009,7 +1009,7 @@ mt7921_tx_complete_status(struct mt76_dev *mdev, struct sk_buff *skb,
 	if (info->flags & IEEE80211_TX_CTL_AMPDU)
 		info->flags |= IEEE80211_TX_STAT_AMPDU;
 
-	if (stat)
+	if (clean_status)
 		ieee80211_tx_info_clear_status(info);
 
 	if (!(info->flags & IEEE80211_TX_CTL_NO_ACK))
@@ -1019,8 +1019,8 @@ mt7921_tx_complete_status(struct mt76_dev *mdev, struct sk_buff *skb,
 	ieee80211_tx_status_ext(hw, &status);
 }
 
-void mt7921_txp_skb_unmap(struct mt76_dev *dev,
-			  struct mt76_txwi_cache *t)
+static void
+mt7921_txp_skb_unmap(struct mt76_dev *dev, struct mt76_txwi_cache *t)
 {
 	struct mt7921_txp_common *txp;
 	int i;
@@ -1050,6 +1050,42 @@ void mt7921_txp_skb_unmap(struct mt76_dev *dev,
 	}
 }
 
+static void
+mt7921_txwi_free(struct mt7921_dev *dev, struct mt76_txwi_cache *t,
+                struct ieee80211_sta *sta, bool clear_status,
+                struct list_head *free_list)
+{
+       struct mt76_dev *mdev = &dev->mt76;
+       struct ieee80211_tx_info *info;
+       __le32 *txwi;
+
+       mt7921_txp_skb_unmap(mdev, t);
+       if (!t->skb)
+               goto out;
+
+       if (!sta)
+               goto out;
+
+       txwi = (__le32 *)mt76_get_txwi_ptr(mdev, t);
+       if (likely(t->skb->protocol != cpu_to_be16(ETH_P_PAE)))
+               mt7921_tx_check_aggr(sta, txwi);
+
+       info = IEEE80211_SKB_CB(t->skb);
+       if (!info->tx_time_est) {
+               struct mt76_wcid *wcid = (struct mt76_wcid *)sta->drv_priv;
+               int pending;
+
+               pending = atomic_dec_return(&wcid->non_aql_packets);
+               if (pending < 0)
+                       atomic_cmpxchg(&wcid->non_aql_packets, pending, 0);
+       }
+
+       mt7921_tx_complete_status(mdev, t->skb, sta, clear_status, free_list);
+out:
+       t->skb = NULL;
+       mt76_put_txwi(mdev, t);
+}
+
 void mt7921_mac_tx_free(struct mt7921_dev *dev, struct sk_buff *skb)
 {
 	struct mt7921_tx_free *free = (struct mt7921_tx_free *)skb->data;
@@ -1108,28 +1144,7 @@ void mt7921_mac_tx_free(struct mt7921_dev *dev, struct sk_buff *skb)
 		if (!txwi)
 			continue;
 
-		mt7921_txp_skb_unmap(mdev, txwi);
-		if (txwi->skb) {
-			struct ieee80211_tx_info *info = IEEE80211_SKB_CB(txwi->skb);
-			void *txwi_ptr = mt76_get_txwi_ptr(mdev, txwi);
-
-			if (likely(txwi->skb->protocol != cpu_to_be16(ETH_P_PAE)))
-				mt7921_tx_check_aggr(sta, txwi_ptr);
-
-			if (sta && !info->tx_time_est) {
-				struct mt76_wcid *wcid = (struct mt76_wcid *)sta->drv_priv;
-				int pending;
-
-				pending = atomic_dec_return(&wcid->non_aql_packets);
-				if (pending < 0)
-					atomic_cmpxchg(&wcid->non_aql_packets, pending, 0);
-			}
-
-			mt7921_tx_complete_status(mdev, txwi->skb, sta, stat, &free_list);
-			txwi->skb = NULL;
-		}
-
-		mt76_put_txwi(mdev, txwi);
+		mt7921_txwi_free(dev, txwi, sta, stat, &free_list);
 	}
 
 	if (wake)
@@ -1307,14 +1322,7 @@ void mt7921_tx_token_put(struct mt7921_dev *dev)
 
 	spin_lock_bh(&dev->mt76.token_lock);
 	idr_for_each_entry(&dev->mt76.token, txwi, id) {
-		mt7921_txp_skb_unmap(&dev->mt76, txwi);
-		if (txwi->skb) {
-			struct ieee80211_hw *hw;
-
-			hw = mt76_tx_status_get_hw(&dev->mt76, txwi->skb);
-			ieee80211_free_txskb(hw, txwi->skb);
-		}
-		mt76_put_txwi(&dev->mt76, txwi);
+		mt7921_txwi_free(dev, txwi, NULL, false, NULL);
 		dev->mt76.token_count--;
 	}
 	spin_unlock_bh(&dev->mt76.token_lock);
diff --git a/drivers/net/wireless/mediatek/mt76/mt7921/mt7921.h b/drivers/net/wireless/mediatek/mt76/mt7921/mt7921.h
index 2d8bd6bfc820..9b701c6b67af 100644
--- a/drivers/net/wireless/mediatek/mt76/mt7921/mt7921.h
+++ b/drivers/net/wireless/mediatek/mt76/mt7921/mt7921.h
@@ -357,8 +357,6 @@ void mt7921_queue_rx_skb(struct mt76_dev *mdev, enum mt76_rxq_id q,
 			 struct sk_buff *skb);
 void mt7921_sta_ps(struct mt76_dev *mdev, struct ieee80211_sta *sta, bool ps);
 void mt7921_stats_work(struct work_struct *work);
-void mt7921_txp_skb_unmap(struct mt76_dev *dev,
-			  struct mt76_txwi_cache *txwi);
 void mt7921_set_stream_he_caps(struct mt7921_phy *phy);
 void mt7921_update_channel(struct mt76_phy *mphy);
 int mt7921_init_debugfs(struct mt7921_dev *dev);
-- 
2.35.1

